main.go

```diff
@@ -2,110 +2,148 @@ package main
+}
+
+type component struct {
+	Producer *producer.Component
+	Postgres *sqldb.PostgresComponent
+	Device42 *ipamfetcher.Device42ClientComponent
+}
+
+func (c *component) Settings() *config {
+	return &config{
+		LambdaMode: false,
+		Producer:   &producerConfig{c.Producer.Settings()},
+		Postgres:   c.Postgres.Settings(),
+		Device42:   c.Device42.Settings(),
+	}
+}
+
+func newComponent() *component {
+	return &component{
+		Producer: producer.NewComponent(),
+		Postgres: sqldb.NewPostgresComponent(),
+		Device42: ipamfetcher.NewDevice42ClientComponent(),
+	}
+}
+
+func (c *component) New(ctx context.Context, conf *config) (func(context.Context, settings.Source) error, error) {
+	p, err := c.Producer.New(ctx, conf.Producer.Config)
 	if err != nil {
-		panic(err.Error())
+		return nil, err
+	}
+	enqueueHandler := &v1.EnqueueHandler{
+		RandomNumberGenerator: &randomnumbergenerator.UUIDGenerator{},
+		Producer:              p,
+		LogFn:                 domain.LoggerFromContext,
 	}
 
-	ipamClient, err := getIPAMClient(ctx, source)
+	pgdb, err := c.Postgres.New(ctx, conf.Postgres)
 	if err != nil {
-		panic(err.Error())
+		return nil, err
 	}
 
-	postgresConfigComponent := &sqldb.PostgresConfigComponent{}
-	postgresdb := new(sqldb.PostgresDB)
-	if err = settings.NewComponent(ctx, source, postgresConfigComponent, postgresdb); err != nil {
-		panic(err.Error())
+	dc, err := c.Device42.New(ctx, conf.Device42)
+	if err != nil {
+		return nil, err
 	}
-	assetFetcher := &assetfetcher.PostgresPhysicalAssetFetcher{DB: postgresdb}
+
+	deviceFetcher := ipamfetcher.NewDevice42DeviceFetcher(dc)
+	subnetFetcher := ipamfetcher.NewDevice42SubnetFetcher(dc)
+	customerFetcher := ipamfetcher.NewDevice42CustomerFetcher(dc)
+	ipamDataFetcher := &ipamfetcher.Client{
+		CustomerFetcher: customerFetcher,
+		DeviceFetcher:   deviceFetcher,
+		SubnetFetcher:   subnetFetcher,
+	}
+
+	assetFetcher := &assetfetcher.PostgresPhysicalAssetFetcher{DB: pgdb}
 	fetchHandler := &v1.FetchByIPAddressHandler{
 		LogFn:                domain.LoggerFromContext,
 		PhysicalAssetFetcher: assetFetcher,
 	}
-	assetStorer := &assetstorer.PostgresPhysicalAssetStorer{DB: postgresdb}
+	assetStorer := &assetstorer.PostgresPhysicalAssetStorer{DB: pgdb}
 	syncHandler := &v1.SyncIPAMDataHandler{
-		IPAMDataFetcher:     ipamClient,
+		IPAMDataFetcher:     ipamDataFetcher,
 		LogFn:               domain.LoggerFromContext,
 		PhysicalAssetStorer: assetStorer,
 	}
+
 	handlers := map[string]serverfull.Function{
 		"fetchbyip": serverfull.NewFunction(fetchHandler.Handle),
 		"sync":      serverfull.NewFunction(syncHandler.Handle),
+		"enqueue":   serverfull.NewFunction(enqueueHandler.Handle),
 	}
 
 	fetcher := &serverfull.StaticFetcher{Functions: handlers}
-	if err := serverfull.Start(ctx, source, fetcher); err != nil {
-		panic(err.Error())
+	if conf.LambdaMode {
+		return func(ctx context.Context, source settings.Source) error {
+			return serverfull.StartLambda(ctx, source, fetcher, "filter")
```

Comment:
Question: what is `filter` in this context?

Reply:
a typo I think. Good catch! :D

Reply:
You may also be the first case where we've both migrated to this kind of component structure at the top level for enabling a Lambda build in addition to having multiple functions within the project. The last argument here is the name of the function that will be served when running on Lambda. I believe you'll need to add another settings to capture the name of the function. When running in Lambda mode someone will need to pass in both `IPAMFACADE_LAMBDAMODE="true"` and `IPAMFACADE_LAMBDAFUNCTION="enqueue"` to have it serve the enqueue feature.

This will enable operators to build the binary once, upload it as a zip to Lambda, and run multiple instances of it with the flag changed in order to host the full service suite of features in AWS Lambda.

Reply:
So, what I'm hearing in what you said is that services with more than one handler will not work at all in Lambda mode, but will be required instead of either run as traditional HTTP services, or as a full Lambda. Is that correct?

Reply:
That depends on how you conceive of Lambda mode versus "full lambda". It's a distinction I don't recognize.

For additional context, switching to Lambda mode changes the startup from using the serverfull HTTP API that implements [Lambda Invoke](https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html) to using the official Lambda SDK for Go which runs using the Go [net/rpc server](https://golang.org/pkg/net/rpc/#Server). The official SDK is only capable of running one function at a time because Lambda, as a platform, only runs one function at a time. In all cases of deploying to Lambda where we want to host multiple functions we have to run them all separately.

Usually, this would require building a binary for each of the various functions we want to upload. Using these flags enables us to use the same binary to do any of these things:

- Run as an HTTP service using the HTTP mode and our API Gateway shim serverfull-gateway
- Run as a compatible Lambda Invoke API with all the contained functions exposed as though they were deployed individually to AWS Lambda
- Run as a native Lambda binary for any selected function in the project

If a project has three functions, we'd still need to upload our binary once for each of the functions we want to run in Lambda in order to give them each a distinct name and configuration. Ultimately, we'd end up running the same code with different configuration and end up with all of our functions running separately.

Reply:
Right, sorry, what I should have said instead of "full lambda" was "without using the API gateway".

We can and should continue this discussion of future options, but not in this PR. 👍